### PR TITLE
Codegen refactoring for operation context params and jmesPath

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
Refactors codegen for operation context parameters and jmesPath code generator, and fixes the condition for when to generate the jmesPath runtime.

## Modifications
- The `EndpointResolverInterceptorSpec` is getting too big. By extracting logic and codegen for Operation Context Params and putting into a separate class, it's easier to read and understand the code
- Renaming the JmesPath code generator from `JmesPathAcceptorGenerator` to `JmesPathExpressionConverter` since `Acceptor` is a Waiter term. Another name could be `JmesPathCodeGenerator`
- During initial development, the conditions for copying the `JmesPathRuntime.java.resource` was set to true. Now that all endpoint parameter logic is finished, it's updated to checking for the presence of Operation Context Params: If there are waiters, we know there is a runtime. If there are no waiters, we'll generate runtime if there are operation context parameters

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import software.amazon.awssdk.codegen.model.rules.endpoints.ParameterModel;
+import software.amazon.awssdk.codegen.poet.PoetExtension;
+import software.amazon.awssdk.codegen.poet.waiters.JmesPathExpressionConverter;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.utils.CollectionUtils;
+
+/**
+ * Knowledge index to get access to operation context parameters.
+ */
+public final class OperationContextParamsKnowledgeIndex {
```
Comment (zoewangg) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java:42]: Nit: `KnowledgeIndex` sounds fancy, but I don't know what it means😅. Maybe just `OperationContextParamsHelper`?

**Review by zoewangg — APPROVED** [review]

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java (hunk 1 above)
Comment (cenedhryn) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java:42]: It's not my favorite either, but there is prior art from the SRA implementation and I'd rather not introduce a new term right now. We could rename all the KnowledgeIndex instances in a separate PR.

Comment (zoewangg) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java:42]: Ah I see, we have more `Helper` than `KnowledgeIndex` though ;) Anyway, it's codegen class, so it doesn't really matter

**Comment by debora-ito** [comment]
Closing stale PRs.

**Pull request closed without merging**